### PR TITLE
[Feature] Implement TypedArray#set with tests

### PR DESCRIPTION
**Pull request by jtenner**

### CODE REVIEW

**Comment by jtenner** [comment]
Tests are probably naive. Could be made better, but it appears to be functional. Any help/guidance would be appreciated.

File: std/assembly/typedarray.ts [hunk 1]
```diff
@@ -1718,3 +1762,41 @@ function WRAP<TArray extends ArrayBufferView, T>(buffer: ArrayBuffer, byteOffset
...
+  let sourceDataStart = source.dataStart;
+
+  // if the types align and match, use memory.copy() instead of manual loop
+  if (isInteger<T>() == isInteger<U>() && alignof<T>() == alignof<U>()) {
+    memory.copy(targetDataStart, sourceDataStart, source.byteLength);
+  } else {
+    let count = source.length;
+    for (let i = 0; i < count; i++) {
+      let targetStoreOffset = <usize>i << alignof<T>();
+      let sourceLoadOffset = <usize>i << alignof<U>();
```
Comment (MaxGraey) [inline std/assembly/typedarray.ts]: `targetStoreOffset` and `sourceLoadOffset` better move inplace (inlined) inside `store`

Comment (jtenner) [reply std/assembly/typedarray.ts]: Won't this just get optimized by binaryen?

Comment (MaxGraey) [reply std/assembly/typedarray.ts]: I guess not necessarily.

File: tests/compiler/std/typedarray.ts [hunk 2]
```diff
@@ -693,3 +693,47 @@ testArrayWrap<Int64Array, i64>();
...
+    if (vala != valb) {
+      trace(nameof<T>(), 3, i, <f64>vala, <f64>valb);
+      assert(false);
+    }
+  }
+}
+
+let setSource1: i32[] = [1, 2, 3];
+let setSource2: f32[] = [4, 5, 6];
+let setSource3: f64[] = [Infinity, NaN, -Infinity];
```
Comment (MaxGraey) [inline tests/compiler/std/typedarray.ts]: I guess need add `i64[]` and typed array sources as well

File: std/assembly/typedarray.ts (hunk 1 above)
Comment (jtenner) [reply std/assembly/typedarray.ts]: The `if` statement is inlined because the conditions are statically known at runtime.

File: tests/compiler/std/typedarray.ts (hunk 2 above)
Comment (jtenner) [reply tests/compiler/std/typedarray.ts]: Okay. I'll add a few more.

File: std/assembly/typedarray.ts (hunk 1 above)
Comment (MaxGraey) [reply std/assembly/typedarray.ts]: I mean use `load(ptr + (i << alignof<T>()))` instead  `load(ptr + cachedStoreOffset)`. That's that case when caching is not helps

Comment (jtenner) [reply std/assembly/typedarray.ts]: Oh that makes a lot of sense. No problem!

File: std/assembly/typedarray.ts [hunk 3]
```diff
@@ -1718,3 +1762,59 @@ function WRAP<TArray extends ArrayBufferView, T>(buffer: ArrayBuffer, byteOffset
...
+              // @ts-ignore: cast to T is valid for numeric types here
+              ? select<T>(0, select<T>(255, <T>value, value > 255), value < 0)
+              : 0
+          );
+        } else {
+          let value = load<U>(sourceDataStart + (<usize>i << alignof<U>()));
+          store<T>(
+            targetDataStart + (<usize>i << alignof<T>()),
+            // @ts-ignore: cast to T is valid for numeric types here
+            select<T>(0, select<T>(255, <T>value, value > 255), value < 0)
```
Comment (MaxGraey) [inline std/assembly/typedarray.ts]: Suggestion:
```ts
if (sizeof<T>() == 4) {
  ...
  ~(<i32>value >> 31) & (((255 - <i32>value) >> 31) | value)
} else {
   ...
   ~(<i64>value >> 63) & (((255 - <i64>value) >> 63) | value)
}
```

File: std/assembly/typedarray.ts [hunk 4]
```diff
@@ -1718,3 +1762,59 @@ function WRAP<TArray extends ArrayBufferView, T>(buffer: ArrayBuffer, byteOffset
...
+      // if TArray is Uint8ClampedArray, then values must be clamped
+      if (target instanceof Uint8ClampedArray) {
+        if (isFloat<U>()) {
+          let value = load<U>(sourceDataStart + (<usize>i << alignof<U>()));
+          store<T>(
+            targetDataStart + (<usize>i << alignof<T>()),
+            isFinite<U>(value)
+              // @ts-ignore: cast to T is valid for numeric types here
+              ? select<T>(0, select<T>(255, <T>value, value > 255), value < 0)
+              : 0
```
Comment (MaxGraey) [inline std/assembly/typedarray.ts]:
```ts
<T>max<U>(0, min<U>(255, value))
```

File: std/assembly/typedarray.ts (hunk 3 above)
Comment (jtenner) [reply std/assembly/typedarray.ts]: This is all set!

File: std/assembly/typedarray.ts (hunk 4 above)
Comment (jtenner) [reply std/assembly/typedarray.ts]: This optimization was a big deal! Thanks for the help.

File: tests/compiler/std/typedarray.ts [hunk 5]
```diff
@@ -693,3 +693,78 @@ testArrayWrap<Int64Array, i64>();
...
+
+function testTypedArraySet<T extends ArrayBufferView>(): void {
+  let setSource4: Int64Array = new Int64Array(3);
```
Comment (MaxGraey) [inline tests/compiler/std/typedarray.ts]:
```suggestion
  let setSource4 = new Int64Array(3);
```

File: tests/compiler/std/typedarray.ts [hunk 6]
```diff
@@ -693,3 +693,78 @@ testArrayWrap<Int64Array, i64>();
...
+
+  let b = new Int64Array(4);
+  b[0] = -10;
+  b[1] = 100;
+  b[2] = 10;
+  b[3] = 300;
+
+  let c = new Int32Array(2);
+  c[0] = 300;
+  c[1] = -1;
```
Comment (MaxGraey) [inline tests/compiler/std/typedarray.ts]: Also add plz `let d = new Uint32Array` and `let e = new Int16Array`

Comment (jtenner) [reply tests/compiler/std/typedarray.ts]: Please check now? Thanks!

Comment (MaxGraey) [reply tests/compiler/std/typedarray.ts]: looks good

**Comment by jtenner** [comment]
Anything else I can do to complete this pull request?

**Comment by MaxGraey** [comment]
Looks good to me

File: std/assembly/typedarray.ts [hunk 7]
```diff
@@ -1718,3 +1762,70 @@ function WRAP<TArray extends ArrayBufferView, T>(buffer: ArrayBuffer, byteOffset
...
+  if (isReference<U>()) {
+    ERROR(E_NOTIMPLEMENTED);
+  }
+
+  // Uncaught RangeError: offset is out of bounds
+  if (offset < 0) throw new RangeError(E_INDEXOUTOFRANGE);
+  if (source.length + offset > target.length) throw new RangeError(E_INDEXOUTOFRANGE);
+
+  // if the types align and match, use memory.copy() instead of manual loop
+  if (isInteger<T>() == isInteger<U>() && alignof<T>() == alignof<U>()) {
```
Comment (dcodeIO) [inline std/assembly/typedarray.ts]: What about floats?

Comment (MaxGraey) [reply std/assembly/typedarray.ts]: Hah, I also asked this. So if both integers equality will true, but if both floats this equality also true (`false == false` => true).

Comment (dcodeIO) [reply std/assembly/typedarray.ts]: Ah, very nice!

File: std/assembly/typedarray.ts [hunk 8]
```diff
@@ -1718,3 +1762,70 @@ function WRAP<TArray extends ArrayBufferView, T>(buffer: ArrayBuffer, byteOffset
...
+      // if TArray is Uint8ClampedArray, then values must be clamped
+      if (target instanceof Uint8ClampedArray) {
+        if (isFloat<U>()) {
+          let value = load<U>(sourceDataStart + (<usize>i << alignof<U>()));
+          store<T>(
+            targetDataStart + (<usize>i << alignof<T>()),
+            isFinite<U>(value) ? <T>max<U>(0, min<U>(255, value)) : 0
+          );
+        } else {
+          let value = load<U>(sourceDataStart + (<usize>i << alignof<U>()));
```
Comment (dcodeIO) [inline std/assembly/typedarray.ts]: Wondering if we'll get more sign-extension/masking here than necessary, and if so whether this can be avoided by typing the intermediate local `i/u32`. Like, if `value` is typed `u16` and we cast that to `i8`, the explicit conversion will wrap, even though the `store8` would already wrap.

Comment (MaxGraey) [reply std/assembly/typedarray.ts]: You mean `store<u8>(<u8>~( .... ))` is unnecessary and just need `load<u8>(~(... ))`?

Comment (dcodeIO) [reply std/assembly/typedarray.ts]: Mostly relevant if `T` and `U` have different sizes, which Wasm has distinct instructions for when loading and storing. The `store<T>` builtin is more of a `store<T>(<U>...)`. It evaluates the argument type in order to pick the right instruction, which implicitly wraps the value. So, if we explicitly convert `<T>someU` that might already be implicit in the store, and any explicit wrapping being unnecessary. Didn't look at the fixtures yet.

Comment (jtenner) [reply std/assembly/typedarray.ts]: Happy to make changes for that. Just want to make sure I understand it before I pull changes from master and make a qualitative change to the code.

Can you help me understand this?

Comment (dcodeIO) [reply std/assembly/typedarray.ts]: I believe what happens in the `T=i8, U=i16` case is that on `value = load<i16>(...), the compiler remembers that `value` is `i16` and wrapped (because the load returns a properly wrapped value). Then, `min<i16>(some, thing)` requires properly wrapped values to determine the minimum, and fortunately it's already wrapped so nothing to do there. After that, the value is cast to `<T>` (`i8` here), which wraps the i16 to an i8 discarding garbage bits (explicit cast does this), but that's unnecessary since what follows is a `store<i8>(...)` translating into an `i32.store8` that already discards garbage bits.

One first thing to test can be to simply remove the `<T>` at the start of the store argument, and see what happens.

Comment (jtenner) [reply std/assembly/typedarray.ts]: image

Wow. @dcodeIO that was an incredible find. A typecast causing an optimization problem like that seems so obscure, I can't believe it.

Anyway, this should be fixed and brought current with master!

File: tests/compiler/std/typedarray.ts (hunk 2 above)
Comment (dcodeIO) [reply tests/compiler/std/typedarray.ts]: Came here looking for an i16->i8 test. Might make sense to have a signed and unsigned small int here, a signed and unsigned normal i32 and a signed and unsigned i64.

File: std/assembly/typedarray.ts [hunk 9]
```diff
@@ -1718,3 +1762,70 @@ function WRAP<TArray extends ArrayBufferView, T>(buffer: ArrayBuffer, byteOffset
...
+          }
+        }
+        // if U is a float, then casting float to int must include a finite check
+      } else if (isFloat<U>() && !isFloat<T>()) {
+        let value = load<U>(sourceDataStart + (<usize>i << alignof<U>()));
+        // @ts-ignore: cast to T is valid for numeric types here
+        store<T>(targetDataStart + (<usize>i << alignof<T>()), isFinite<U>(value) ? <T>value : 0);
+      } else {
+        // @ts-ignore: cast to T is valid for numeric types here
+        store<T>(targetDataStart + (<usize>i << alignof<T>()), <T>load<U>(sourceDataStart + (<usize>i << alignof<U>())));
```
Comment (dcodeIO) [inline std/assembly/typedarray.ts]: Looks like this is the int->int case, in which case the `<T>` is unnecessary for the same reason. I believe the tests currently do not catch this due to missing small and long int cases.

Comment (jtenner) [reply std/assembly/typedarray.ts]:
```
trace("hit!");
store<T>(targetDataStart + (<usize>i << alignof<T>()), <T>load<U>(sourceDataStart + (<usize>i << alignof<U>())));
```

This is the line in question yes?

```
- instantiate untouched OK (33.971 ms)
  716 allocs, 716 frees, 2077 increments, 2077 decrements

  [fun] __start = function 674() { [native code] }
  [obj] memory = [object WebAssembly.Memory]

  [start]
  trace: Hit!
  trace: Hit!
  trace: Hit!
...
```

This line is hit quite regularly. I will simply remove the offending `<T>` and see what happens.

Comment (dcodeIO) [reply std/assembly/typedarray.ts]: Is it being hit with small int types for both T and U?

```ts
if (sizeof<U>() < 4) trace("hit U!");
if (sizeof<Z>() < 4) trace("hit Z!");
```

Comment (jtenner) [reply std/assembly/typedarray.ts]: Okay. I have evidence that this optimization isn't too necessary, but it does seem to help.

image

This is only in the `untouched` version, and the optimized version remains unchanged.

I had to change it so that the cast happens if the conversion is from integer to float, however, by adding another `else if` branch to take advantage of this optimization.

Comment (dcodeIO) [reply std/assembly/typedarray.ts]: The `i64` to native `i32` (or smaller) and vice-versa cases are interesting for the optimizer because it can potentially remain on i32 or i64 exclusively, without a `.wrap_i64` instruction for example.

Other interesting cases are `i8` -> `i32`, `u8` -> `i16`, `u16` -> `i8` because these involve sign extension/wrapping. That's what I'm trying to get to, as these aren't currently present in the tests.

File: tests/compiler/std/typedarray.ts (hunk 2 above)
Comment (jtenner) [reply tests/compiler/std/typedarray.ts]: Please forgive me for finally getting around to this. I added a few more tests and actually found an edge case involving `Uint8ClampedArray`.

I would be happy to add some more tests if you'd like! I just need to get around to them tomorrow.

Can I get some feedback on the ones I put in?

File: std/assembly/typedarray.ts (hunk 8 above)
Comment (jtenner) [reply std/assembly/typedarray.ts]: I removed the offending `<T>`, and this should at least be resolved from an implementation standpoint, if not from the testing standpoint. See above!

Can we mark this conversation as resolved?

File: std/assembly/typedarray.ts (hunk 9 above)
Comment (jtenner) [reply std/assembly/typedarray.ts]: Okay. I'll talk with @MaxGraey to get the test cases all covered.

File: tests/compiler/std/typedarray.ts (hunk 2 above)
Comment (dcodeIO) [reply tests/compiler/std/typedarray.ts]: Looks good now with the additional Uint8Array and Int16Array cases.
